Lib/importlib/_bootstrap_external.py

```diff
@@ -1711,6 +1711,78 @@ def __repr__(self):
         return f'FileFinder({self.path!r})'
 
 
+class AppleFrameworkLoader(ExtensionFileLoader):
+    """A loader for modules that have been packaged as frameworks for
+    compatibility with Apple's iOS App Store policies.
+
+    For compatibility with the iOS App Store, *all* binary modules in an iOS
+    app must be .dylib objects, contained in a framework with appropriate
+    metadata, stored in the ``Frameworks`` folder of the packaged app. There
+    can be only a single binary per framework, and there can be no executable
+    binary material outside the Frameworks folder.
+
+    If you're trying to run ``from foo.bar import _whiz``, and ``_whiz`` is
+    implemented with the binary module ``foo/bar/_whiz.abi3.dylib`` (or any
+    other ABI .dylib extension), this loader will look for
+    ``{sys.executable}/Frameworks/foo.bar._whiz.framework/_whiz.abi3.dylib``
+    (forming the package name by taking the full import path of the library,
+    and replacing ``/`` with ``.``).
+
+    However, the ``__file__`` attribute of the ``_whiz`` module will report as
+    the original location inside the ``foo/bar`` subdirectory. This so that
+    code that depends on walking directory trees will continue to work as
+    expected based on the *original* file location.
+
+    The Xcode project building the app is responsible for converting any
+    ``.dylib`` files from wherever they exist in the ``PYTHONPATH`` into
+    frameworks in the ``Frameworks`` folder (including the addition of
+    framework metadata, and signing the resulting framework). This will usually
+    be done with a build step in the Xcode project; see the iOS documentation
+    for details on how to construct this build step.
+    """
+    def __init__(self, fullname, dylib_file, parent_paths=None):
+        super().__init__(fullname, dylib_file)
+        self.parent_paths = parent_paths
+
+    def create_module(self, spec):
+        mod = super().create_module(spec)
+        if self.parent_paths:
+            for parent_path in self.parent_paths:
+                if _path_isdir(parent_path):
+                    mod.__file__ = _path_join(
+                        parent_path,
+                        _path_split(self.path)[-1],
+                    )
+                    break
+        return mod
+
+
+class AppleFrameworkFinder:
+    """A finder for modules that have been packaged as Apple Frameworks
+    for compatibility with Apple's App Store policies.
+
+    See AppleFrameworkLoader for details.
+    """
```

Comment:
> It would also mean that it would be checked (unnecessarily) for any other `sys.path` entries

Out of curiosity, what should one expect `sys.path` to look like on iOS?

Also, if I've understood correctly then extension modules can only be found under the app's frameworks folder and never on any other `sys.path` entry.  If that's the case then should we remove the `ExtensionFileLoader` entry from `_get_supported_file_loaders()` in Lib/importlib/_bootstrap_external.py?  Also, are the restrictions on extension modules (location, suffix) technical restrictions or strictly a consequence of the fact that apps cannot ship with non-conforming binary files, as enforced by the app store police?  Is there a possibility that one could find an extension file out of place somehow, like in some system directory?

IIUC, .py files may still be imported using the normal path-entry finder (`FileFinder`), right?  I'm not familiar with iOS apps to even imagine when the finder might be looking (i.e. what's in `sys.path`).

Relatedly, what happens if an app maintainer (or, somehow, a user) registers additional importers or messes with `sys.path`?  Could that cause problems for this finder?  Could it circumvent the app store rules?

> and that any other path-entry finders would check the framework-specific `sys.path` entries unnecessarily.

FYI, the import machinery already makes this (almost) a non-issue with path hooks and `sys.path_importer_cache`.

Reply:
You're correct that the reason for the metapath finder is the structure of the Frameworks folder. I hadn't considered using a custom FileFinder; I agree we should do anything we can to avoid breaking user expectations (however eccentric or unexpected those might be), so I'll take a look and see if I can make this work.

As for ExtensionFileLoader - that's an interesting case. iOS apps are entirely *capable* of loading binary modules from the PYTHONPATH; it's entirely an issue of the resulting app being acceptable to the App Store. If you've got a project that *doesn't* have the "move dylibs to the Frameworks folder" build step, the app will still *run* - it will just be rejected when you try to submit it to the App store (and it will be rejected by an automated check at time of submission, not hours/days/weeks later as a result of the review process). I've also had a couple of bugs (including one I'm still chasing with cryptography) where it's useful to be able to confirm if the problem you're seeing is because the .dylib has been moved, or because the .dylib isn't working. I therefore opted to retain the ExtensionFileLoader, just in case it was useful. 

The SourceFileLoader and SourcelessFileLoader both work exactly the same, however. Loading `.py` files and `.pyc` files from anywhere in the app bundle isn't an issue; it's only binary modules that the App Store restricts. `sys.path` for a running app will contain `{dirname(sys.executable)}/python/lib/python3.X` at a minimum (i.e., `prefix=={dirname(sys.executable)}/python`, but the bin/include/man/share folders aren't installed). BeeWare apps usually end up with `{dirname(sys.executable)}/app` and `{dirname(sys.executable)}/app_packages` as well as the location of app code and app dependencies, but that's entirely up to the app itself. 

If a user installs an additional importer... I guess that depends on what the importer is doing. The entire contents of the app bundle is fair game for reading; so a novel mechanism for finding .py files shouldn't be an issue. The only place I could force a problem is if the importer is expecting to find an executable binary in a specific location outside the Frameworks folder - but the app won't be accepted into the App Store in the first place if they try to do this. As for circumventing the App Store rules - it might be a lack of imagination on my part, but I'm having difficulty imaging what you'd be able to do here. Normal filesystem write protections should prevent even a malicious user from doing anything damaging, and you won't be able to use anything binary that isn't in the app bundle and signed on that basis.

Reply:
Sounds good.  Thanks for the detailed explanation!